dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java

```diff
@@ -277,65 +284,70 @@ private BooleanFormula encodeForwardProgress(Program program, ProgressModel prog
     public BooleanFormula encodeMemory() {
         logger.info("Encoding memory");
-        final Memory memory = context.getTask().getProgram().getMemory();
-        final FormulaManager fmgr = context.getFormulaManager();
-        // TODO: Once we want to encode dynamic memory layouts (e.g., due to dynamically-sized mallocs)
-        //  we need to compute a Map<MemoryObject, Formula> where each formula describes a
-        //  unique, properly aligned, and non-overlapping address of the memory object.
-        final Map<MemoryObject, BigInteger> memObj2Addr = computeStaticMemoryLayout(memory);
-
-        final var enc = new ArrayList<BooleanFormula>();
-        for (final MemoryObject memObj : memory.getObjects()) {
-            // For all objects, their 'final' value fetched here represents their constant value.
-            final Formula addressVariable = context.encodeFinalExpression(memObj);
-            final BigInteger addressInteger = memObj2Addr.get(memObj);
-
-            if (addressVariable instanceof BitvectorFormula bitvectorVariable) {
-                final BitvectorFormulaManager bvmgr = fmgr.getBitvectorFormulaManager();
-                final int length = bvmgr.getLength(bitvectorVariable);
-                enc.add(bvmgr.equal(bitvectorVariable, bvmgr.makeBitvector(length, addressInteger)));
+        return encodeMemoryLayout(context.getTask().getProgram().getMemory());
+    }
+
+    private BooleanFormula encodeMemoryLayout(Memory memory) {
+        final BooleanFormulaManager bmgr = context.getBooleanFormulaManager();
+        final EncodingHelper helper = new EncodingHelper(context.getFormulaManager());
+        final List<BooleanFormula> enc = new ArrayList<>();
+
+        // TODO: We could sort the objects to generate better encoding:
+        //  "static -> dynamic with known size -> dynamic with unknown size"
+        //  For the former two we can then statically assign addresses as we do now and
+        //  only the latter needs dynamic encodings.
+        final List<MemoryObject> memoryObjects = ImmutableList.copyOf(memory.getObjects());
+        for (int i = 0; i < memoryObjects.size(); i++) {
+            final MemoryObject cur = memoryObjects.get(i);
+            final Formula addr = context.address(cur);
+            final Formula size = context.size(cur);
+            final Formula alignment;
+
+            // Encode size & compute alignment
+            if (cur.isStaticallyAllocated()) {
+                enc.add(helper.equals(size, context.encodeFinalExpression(cur.size())));
+                alignment = context.encodeFinalExpression(cur.alignment());
             } else {
-                assert addressVariable instanceof IntegerFormula;
-                final IntegerFormulaManager imgr = fmgr.getIntegerFormulaManager();
-                enc.add(imgr.equal((IntegerFormula) addressVariable, imgr.makeNumber(addressInteger)));
+                // Non-allocated objects get size 0
```

Comment:
Please add "i.e. objects whose corresponding alloc event was not executed", It took me a while to understand what "non-allocated mean"